File: problems/tree/0543-diameter-of-binary-tree.js
/**
 * #543 Diameter of Binary Tree
 * Each recursive call returns a subtree height; combine child heights at a node.
 * The diameter is measured in edges, so a through-node path is left + right.
 * Time: O(n) | Space: O(h) call stack
 *
 * @param {TreeNode|null} root
 * @return {number}
 */
function diameterOfBinaryTree(root) {
  let diameter = 0

  function height(node) {
    if (node === null) return 0

    const leftHeight = height(node.left)
    const rightHeight = height(node.right)
    diameter = Math.max(diameter, leftHeight + rightHeight)
    return 1 + Math.max(leftHeight, rightHeight)
  }

  height(root)
  return diameter
}

module.exports = diameterOfBinaryTree

File: problems/tree/0543-diameter-of-binary-tree.py
"""#543 Diameter of Binary Tree
Return subtree height; at each node, left height plus right height is a path.
Time: O(n) | Space: O(h) call stack.
"""


class Solution:
    def diameterOfBinaryTree(self, root):
        diameter = 0

        def height(node):
            nonlocal diameter
            if node is None:
                return 0
            left_height = height(node.left)
            right_height = height(node.right)
            diameter = max(diameter, left_height + right_height)
            return 1 + max(left_height, right_height)

        height(root)
        return diameter
